lib/android.js

```diff
@@ -1498,54 +1659,534 @@ function unwrapMethodId (methodId) {
+
+    if ((originalFlags & kAccXposedHookedMethod) !== 0 && xposedIsSupported()) {
       const hookInfo = this.originalMethod.jniCode;
-      this.hookedMethod = hookInfo.add(2 * pointerSize).readPointer();
-      this.originalMethod = fetchArtMethod(this.hookedMethod, vm);
+      this.hookedMethodId = hookInfo.add(2 * pointerSize).readPointer();
+      this.originalMethod = fetchArtMethod(this.hookedMethodId, vm);
     }
 
-    const { hookedMethod } = this;
-
-    const originalFlags = hookedMethod.add(artMethodOffset.accessFlags).readU32();
+    const { hookedMethodId } = this;
 
-    // kAccFastNative so that the VM doesn't get suspended while executing JNI
-    // (so that we can modify the ArtMethod on the fly)
-    // Remove kAccFastInterpreterToInterpreterInvoke to disable use_fast_path in interpreter_common.h
-    const replacementFlags = ((originalFlags & ~(kAccCriticalNative | kAccFastInterpreterToInterpreterInvoke)) |
-        kAccNative | kAccFastNative) >>> 0;
+    const replacementMethodId = cloneArtMethod(hookedMethodId, vm);
+    this.replacementMethodId = replacementMethodId;
 
-    patchArtMethod(hookedMethod, {
+    patchArtMethod(replacementMethodId, {
       jniCode: impl,
-      accessFlags: replacementFlags,
-      quickCode: api.artQuickGenericJniTrampoline,
+      accessFlags: ((originalFlags & ~(kAccCriticalNative | kAccFastNative)) | kAccNative) >>> 0,
+      quickCode: api.artClassLinker.quickGenericJniTrampoline,
       interpreterCode: api.artInterpreterToCompiledCodeBridge
     }, vm);
 
-    notifyArtMethodHooked(hookedMethod);
+    // Remove kAccFastInterpreterToInterpreterInvoke and kAccSkipAccessChecks to disable use_fast_path
+    // in interpreter_common.h
+    let hookedMethodRemovedFlags = kAccFastInterpreterToInterpreterInvoke;
+    if ((originalFlags & kAccNative) === 0) {
+      hookedMethodRemovedFlags |= kAccSkipAccessChecks;
+    }
+
+    patchArtMethod(hookedMethodId, {
+      accessFlags: (originalFlags & ~(hookedMethodRemovedFlags)) >>> 0
+    }, vm);
+
+    const quickCode = this.originalMethod.quickCode;
+
+    // Replace Nterp quick entrypoints with art_quick_to_interpreter_bridge to force stepping out
+    // of ART's next-generation interpreter and use the quick stub instead.
+    if (api.artNterpEntryPoint !== undefined && quickCode.equals(api.artNterpEntryPoint)) {
+      patchArtMethod(hookedMethodId, {
+        quickCode: api.artQuickToInterpreterBridge
+      }, vm);
+    }
+
+    if (!isArtQuickEntrypoint(quickCode)) {
+      this.interceptor = new ArtQuickCodeInterceptor(quickCode);
+      this.interceptor.activate(vm);
+    }
+
+    replacedMethods.set(hookedMethodId.toString(), replacementMethodId);
+
+    notifyArtMethodHooked(hookedMethodId, vm);
   }
 
   revert (vm) {
-    patchArtMethod(this.hookedMethod, this.originalMethod, vm);
+    patchArtMethod(this.hookedMethodId, this.originalMethod, vm);
+
+    replacedMethods.delete(this.hookedMethodId.toString());
```

Comment:
Nit: `this.hookedMethodId` is looked up twice. Could use destructuring to read it and `interceptor` in one go.